Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift

```diff
@@ -18,43 +18,177 @@ struct UnusedEnumeratedRule: Rule {
+                onePosition: firstTokenIsUnderscore ? nil : secondElement.positionAfterSkippingLeadingTrivia
+            )
+        }
+
+        override func visit(_ node: FunctionCallExprSyntax) -> SyntaxVisitorContinueKind {
+            guard node.isEnumerated,
+                  let parent = node.parent,
+                  parent.as(MemberAccessExprSyntax.self)?.declName.baseName.text != "filter",
+                  let trailingClosure = parent.parent?.as(FunctionCallExprSyntax.self)?.trailingClosure
+            else {
+                return .visitChildren
+            }
+
+            if let parameterClause = trailingClosure.signature?.parameterClause {
+                guard let parameterClause = parameterClause.as(ClosureShorthandParameterListSyntax.self),
+                      parameterClause.count == 2,
+                      let firstElement = parameterClause.first,
+                      let secondElement = parameterClause.last
+                else {
+                    return .visitChildren
+                }
+
+                let firstTokenIsUnderscore = firstElement.isUnderscore
+                let lastTokenIsUnderscore = secondElement.isUnderscore
+                guard firstTokenIsUnderscore || lastTokenIsUnderscore else {
+                    return .visitChildren
+                }
+
+                addViolation(
+                    zeroPosition: firstTokenIsUnderscore ? firstElement.positionAfterSkippingLeadingTrivia : nil,
+                    onePosition: firstTokenIsUnderscore ? nil : secondElement.positionAfterSkippingLeadingTrivia
+                )
             } else {
-                position = secondElement.positionAfterSkippingLeadingTrivia
-                reason = "When the item is not used, `.indices` should be used instead of `.enumerated()`"
+                nextClosureId = trailingClosure.id
             }
 
-            violations.append(ReasonedRuleViolation(position: position, reason: reason))
+            return .visitChildren
+        }
+
+        override func visit(_ node: ClosureExprSyntax) -> SyntaxVisitorContinueKind {
+            if let nextClosureId, nextClosureId == node.id {
+                closures.push(Closure(id: nextClosureId))
+                self.nextClosureId = nil
+            }
+            return .visitChildren
+        }
+
+        override func visitPost(_ node: ClosureExprSyntax) {
+            guard let closure = closures.peek(), node.id == closure.id else {
+                return
+            }
+            defer { closures.pop() }
+
+            guard (closure.zeroPosition != nil) != (closure.onePosition != nil) else {
+                return
+            }
+
+            addViolation(zeroPosition: closure.zeroPosition, onePosition: closure.onePosition)
+        }
+
+        override func visitPost(_ node: DeclReferenceExprSyntax) {
+            guard var closure = closures.peek(), node.baseName.text == "$0" || node.baseName.text == "$1" else {
```

Comment:
> How does it make sure that the closure is the right one?

So when we entered the closure, in `visit(_ node: ClosureExprSyntax)`, if the closure was the trailing closure of an enumerated function call (whose id is stored in `nextClosureId`) we push it onto the stack, so that is what `closures.peek()` will return.

And when we leave the closure, in `visitPost(_ node: ClosureExprSyntax)`, we pop that closure off the stack, and update state for that enclosing closure.

Reply:
Consider
```swift
list.enumerated().map {
    $1.forEach { print($0) }
    return $1 
}
```
which is expected to trigger on the last `$1`, right? But does it actually? 🤓

Reply:
Dammit. I think I might need to go back to the drawing board on this one :-(

Reply:
Hint: `ShorthandArgumentRule`